imitation/model_env.py

```diff
@@ -0,0 +1,345 @@
+"""Finite-horizon discrete environments with known transition dynamics. These
+are handy when you want to perform exact maxent policy optimisation."""
+
+import abc
+
+import gym
+from gym import spaces
+import numpy as np
+
+
+class ModelBasedEnv(gym.Env, abc.ABC):
+    """ABC for tabular environments with known dynamics."""
+    def __init__(self):
+        self.cur_state = None
+        self.n_actions_taken = None
+        # we must cache action & observation spaces instead of reconstructing
+        # anew so that the random state of the action space (!!) is preserved
+        self._action_space = None
+        self._observation_space = None
+        self.seed()
+
+    @property
+    def action_space(self):
+        if not self._action_space:
+            self._action_space = spaces.Discrete(self.n_actions)
+        return self._action_space
+
+    @property
+    def observation_space(self):
+        if not self._observation_space:
```

Comment:
For better or worse, it is standard Gym interface to *not* have `seed()` seed the observation and action spaces.

The benefit of having a property is that the `observation_space` is not mutable by external users.

I don't see why we have the `if not self._observation_space` logic though: why not just always set it in the constructor?

Reply:
I did that because `.observation_space` requires `.obs_dim` to construct, which in turn requires `.observation_matrix` (in the default implementation, at least). Setting up `.observation_matrix` is the job of the subclass, though, so if the subclass calls `super().__init__()` before doing anything else, then it won't be possible to set up `.observation_space` in the `ModelBasedEnv` constructor.

There may be a less hacky way to do this. I'll talk with Steven.

Reply:
Edit: talked to Steven & decided it's not worth changing. Here's a new comment explaining why I'm doing that:

```
        # Constructing action & observation spaces requires self.n_actions and
        # self.obs_dim, which are set in subclasses. If we constructed
        # observation & action in this __init__ method, then subclasses would
        # have to call super().__init__() last to give it access to
        # obs_dim/n_actions. By constructing these lazily, we ensure that
        # subclasses can call super().__init__() at any point & still have it
        # succeed.
```